Support IPv4 and IPv6 with prefix length, using a Trie

thebracket opened the pull request:
This PR does the following:

* Replaced the `map_ip_hash` structure with a `BPF_MAP_TYPE_LPM_TRIE`. The key is a new `ip_hash_key` structure, which contains a prefix length and 128-bit address storage.
* The `xdp_iphash_to_cpu_cmdline` has gained an extra flag, `--prefix`. You can add an address with a /24 using `xdp_iphash_to_cpu_cmdline --add --ip 100.64.1.3 --cpu 0 --classid 1:5 --prefix 24`.
* If you don't include `--prefix`, it defaults to `/32` for IPv4 and `/128` for IPv6 --- preserving compatibility with the existing setup (no changes required if you don't want to use these features).
* The command-line also now understands IPv6 addresses. So `xdp_iphash_to_cpu_cmdline --add --ip fe80::215:5dff:fe64:4d05 --cpu 0 --classid 1:5 --prefix 64` will setup a matcher for that IPv6 address and all of its children.
* The command-line `--list` function has changed to also show the prefix, e.g. `"100.64.1.3/24" : { "cpu" : 0, "tc_maj" : "1" , "tc_min" : "5" },` or `"fe80::215:5dff:fe64:4d05/64" : { "cpu" : 0, "tc_maj" : "1" , "tc_min" : "5" }`
* The `clear` command no-longer takes the really odd route of translating an address into a string, and then translating it right back again for each address to remove it; it just removes all items.
* Once read, all addresses are turned into 128-bit addresses. IPv4 is placed in the last 4 octets of the address, with the rest zeroed. The command-line adds 96 to IPv4 prefixes automatically, so you don't need to remember to turn /24 into /120 (which it will be internally).
* The classifier and `xdp_iphash_to_cpu` program use a Trie lookup, starting with 128-bits (trying for an exact match). In testing, it seems very fast. (I haven't formally benchmarked it)

My test environment consists of 3 VMs: one running the shaper (I'm using LibreQOS to generate the basic commands), and a VM on either side of a bridge in the shaper. `iperf` shows that it is shaping correctly both for known and unknown addresses, and is correctly placing prefixed addresses in the right group.

Also, if you add this you may want to reboot to get rid of the previous pinned map. I didn't add any code to remove/replace it.

FIXES https://github.com/xdp-project/xdp-cpumap-tc/issues/4

Replaces https://github.com/xdp-project/xdp-cpumap-tc/pull/13, with the addition of sign-offs and breaking it into separate commits.

References: https://github.com/rchac/LibreQoS/issues/57
Tagging: @tohojo @interduo @rchac

thebracket committed:
shared_maps.h, common_kern_user.h: Change map_ip_hash to a BPF_MAP_TYPE_LPM_TRIE

It's very common in IPv4 networks to need to apply a shaper rule to a subnet, rather than an
individual host. For example, if a customer occupies the entire 192.168.100.0/24 subnet - but
the provider is selling a single 100 Mbps connection - the entire subnet will likely fall
inside a single classifier rule. Previous versions of xdp-cpumap-tc would require the creation
of 254 entries in map_ip_hash to accomplish this - quickly exhausting the 32k rules limit for
a large Internet Service Provider.

The problem becomes more extreme when IPv6 is employed. If RIPE's recommendations [1] are
applied, each customer receives a /64 allocation. Creating 2^64 map entries per customer
is impractical (and won't work at all with IP_HASH_ENTRIES set to 32,767).

In 2017, BPF_MAP_TYPE_LPM_TRIE was added to the Linux BPF implementation to avoid this problem.[2]
BPF_MAP_TYPE_LPM_TRIE implements a "longest prefix match" unbalanced Trie: keys are stored as a
combined "prefix length" and address. For example, 192.168.100.0/24 would be stored as:

* prefix: 24
* address: 192.168.100.0

Longest Prefix Matching (LPM) ensures that the most-specific match is selected, falling back to
larger subnets/smaller prefixes when a specific match is not found. For example, 192.168.100.5 would
match 192.168.100.0/24 in the absence of a more-specific key.

This commit creates a new type, named ip_hash_key, to replace the __u32 specific addresses
used in previous versions. The key contains a prefix (which must be in the range 0 through 128)
and 128-bits of address, in the form of an IPv6 address. IPv4 addresses should be stored in the
last 32-bits, with the preceeding bits zeroed. Queries should be performed with a starting prefix
of 128, ensuring the most-specific IPv4 or IPv6 match is returned.

The new map style reduces storage requirements for an IPv4 /24 block from 254 entries to
1. Likewise, an IPv6 /64 requires 1 map entry rather than 18,446,744,073,709,551,616.

If you are classifying subnets, this represents a significant improvement in memory utilization.
If you are only classifying individual (/32) IPv4 addresses, then the new approach adds overhead:
each address now consumes 160 bits (128-bit address, 32-bit prefix) rather than 32.

Using a Trie as opposed to a map also imposes a small performance penalty when compared with
individual address lookups. Directly hashing a 32-bit address (into a 32-bit lookup value)
is faster than tree-walking. This is a trade-off for the flexibility of being able to classify
an entire subnet (or IPv6 address) with a single map entry.

[1] - https://www.ripe.net/publications/docs/ripe-690
[2] - https://github.com/torvalds/linux/commit/b95a5c4db09bc7c253636cb84dc9b12c577fd5a0

Signed-off-by: Herbert Wolverson <herberticus@gmail.com>

thebracket committed:
tc_classify_kern.c: Use the new LPM_TRIE structure for lookups, add IPv6 support.

This commit imports IPv6 parsing code from the IPv6 branch, allowing the BPF
program to identify both IPv6 and IPv4 packets. Packet addresses are encoded
into an ip_hash_key structure (the key type for the new Trie map):

* IPv4 addresses are stored in the last 32-bits of the key address,
  with the preceding entries zeroed.
* IPv6 address occupy the entire address portion of the key.

Once the address has been encoded in an ip_hash_key structure, a LPM
(Longest Prefix Match) search is performed with a prefix length of 128,
ensuring a match with the most-specific member of mapfile_ip_hash.

Signed-off-by: Herbert Wolverson <herberticus@gmail.com>

thebracket committed:
xdp_iphash_to_cpu_kern.c: Replace parse_ipv4 with parse_ip

Replaces the parse_ipv4 function with a new parse_ip function (derived
from a similar function in the IPv6 branch of this repository). The
new function reads the eth_proto of each packet, and extracts IPv4 or
IPv6 addresses into an ip_hash_key structure. IPv4 addresses are packed
into the last 32-bits of the structure's address (with the preceding
items zeroed), IPv6 addresses consume the entire lookup address.

Once the address has been parsed, a lookup (starting with a prefix of
128, using longest prefix matching to ensure a most-specific match) is
used to return the mapping information for an IP address. The program
then continues as it did before - assigning a CPU to the incoming
packet.

Signed-off-by: Herbert Wolverson <herberticus@gmail.com>

thebracket committed:
common_user.h/common_user.c: Update IP address parsing.

Adds a helper function (ip_string_to_key) to convert IP address strings
to ip_hash_key structures. The function uses getaddrinfo to parse
user input and determine whether an address was an IPv4, IPv6 or invalid
address.

* IPv4 addresses are packed into the last 32-bits of the lookup address,
  with the preceding entries zeroed.
* IPv6 address occupy the key's entire address field.
* Invalid addresses return a prefix-length of 0 to indicate invalid.

Prefix length is stored as either the provided prefix + 96 for IPv4 (to
account for the 128-bit storage scheme) or as specified for IPv6.

Adds a "prefix" field to the iphash_modify function, which is passed to
the ip_string_to_key function. iphash_modify is adjusted to return
IP_FAIL if ip_string_to_key returns a prefix length of 0, indicating
an invalid address.

Signed-off-by: Herbert Wolverson <herberticus@gmail.com>

thebracket committed:
xdp_iphash_to_cpu_cmdline.c: Adjust IP string support to use the new syntax.

Updates the CLI interface to support using --prefix to specify an IP address
prefix length. For example: xdp_iphash_to_cpu_cmdline --add --ip 1.2.3.4 --prefix 24

Updates handling of map modification to pass the provided prefix to ip_hash_modify.

Replaces handling of --clear with a simple removal, rather than converting each address
into a string and then back again.

Signed-off-by: Herbert Wolverson <herberticus@gmail.com>

thebracket committed:
xdp_iphash_to_cpu_cmdline.c: Remove -6 option, included by mistake.

The previous commit accidentally included an option from early development
that is no longer needed.

Signed-off-by: Herbert Wolverson <herberticus@gmail.com>

thebracket committed:
common_user.c, common_user.h, xdp_iphash_to_cpu_cmdline.c: Add direct parsing of CIDR addresses

Updates the ip_string_to_key function to search the provided IP address string
for a '/'. If one is provided, the prefix length is extracted from the subsequent
digits. If no prefix is provided, default to etiher /32 (IPv4) or /128 (IPv6).

Removes the requirement for a prefix parameter for ip_string_to_key and
ip_hash_modify, and restores the functions' original headers. Adjusts the
calls to these functions in xdp_ip_hash_to_cpu_cmdline.c.

Removes the --prefix CLI option, since it is no longer needed.

Signed-off-by: Herbert Wolverson <herberticus@gmail.com>

thebracket committed:
src/common_user.c: use INET6_ADDRSTRLEN for consistency.

For consistency, use INET6_ADDRSTRLEN rather than a literal for IP string buffer length.

Signed-off-by: Herbert Wolverson <herberticus@gmail.com>

netoptimizer commented on src/tc_classify_kern.c:
```diff
@@ -177,8 +181,35 @@ __u32 get_ipv4_addr(struct __sk_buff *skb, __u32 l3_offset, __u32 ifindex_type)
+
+	if (ip6h + 1 > data_end) {
+		//bpf_debug("Invalid IPv6 packet: L3off:%llu\n", l3_offset);
+		//*dst = nulladdr;
```
What is this leftover code in the comment "nulladdr".
Please remove that comment

interduo replied:
Resolved by https://github.com/xdp-project/xdp-cpumap-tc/pull/14/commits/ac81223309022e7375b6ae341c92ca79ebefc314

netoptimizer reviewed (changes requested):
This commit ("Change map_ip_hash to a BPF_MAP_TYPE_LPM_TRIE [...]") description need to argue/describe why it makes sense to change to the LPM map type.

And mention that LPM stands for Longest Prefix Match.

netoptimizer commented on src/xdp_iphash_to_cpu_kern.c:
```diff
@@ -102,55 +102,95 @@ bool parse_eth(struct ethhdr *eth, void *data_end,
+		 * contacting captive portal (which are not yet configured)
+		 */
+		// bpf_debug("cant find ip_info->cpu id for ip:%u\n", ip);
+		// the all-zeroes address is for default traffic
```
Here the code change the missed lookup "unknown traffic" default entry from `255.255.255.255` to `0.0.0.0`

interduo replied:
Resolved by https://github.com/xdp-project/xdp-cpumap-tc/pull/14/commits/085be76ec96f1fd9f114817d47435e9ec2c6f790

thebracket committed:
tc_classify_kern.c: Remove dangling comment

Removes a commented-out line of code left-over from merging from
the IPv6 branch.

Signed-off-by: Herbert Wolverson <herberticus@gmail.com>

thebracket committed:
common_user.c: Replace 0 with 255 as the error-code for failing to parse an IP address.

ip_string_to_key indicates success or failure with the prefixlen of the
returned ip_hash_key. Previously, it returned 0 to indicate failure.
This would collide with deliberately using a prefix of 0 (for a default
match), preventing 0.0.0.0/0 from being added to the lookup map.

The error code is now 255, which should never occur for an IPv4 or
IPv6 address.

Signed-off-by: Herbert Wolverson <herberticus@gmail.com>

thebracket committed:
Change IPv4 packing to a 0xFFFFFFFF prefix from 0x0.

Previous versions packed IPv4 keys as 96 bits of zeroes, followed by the
32-bit IPv4 address. While this worked, it caused some issues:

* Specifying 0.0.0.0/0 as a default handler resulted in the lister
  deducing that the address was an IPv4 address and displaying
  the incorrect value.
* Current users relied on matching 255.255.255.255 for default
  queue handling. This was broken in this patch.

This commit replaces the zero padding with all ones. So an
ip_hash_key containing an IPv4 address is now represented as:
   0xFFFFFF 0xFFFFFFFF 0xFFFFFFFF (the address)

All default setup of ip_hash_key structures are replaced with
the new scheme in this patch.

Handling of 255.255.255.255 as a default is re-enabled in
both xdp_iphash_to_cpu_kern.c and tc_classify_kern.c. This
imposes a small overhead (a second map lookup), but preserves
compatibility.

A preferred approach is to specify "::/0" as a default address
(an IPv6 default route). This will match any provided address/
prefix combination, and not incur the cost of a secondary route
lookup.

Tested with:

* 0.0.0.0/0 as a default, which correctly matches IPv4 as a default.
  It does NOT match unspecified IPv6 addresses. You can use this
  if you desire alternative behavior for un-matched IPv4 vs IPv6.
* ::/0 as a default, correctly matched both IPv4 and IPv6. You
  can use this to match any address.
* 255.255.255.255 (/32 - not specified), which correctly applied the
  default to IPv4 addresses. Will not trigger a match for IPv6
  addresses.

Signed-off-by: Herbert Wolverson <herberticus@gmail.com>

netoptimizer reviewed (approved):
Looked though all of the commits

- they have nice descriptions now
- the code also look nice

I like that you came with a scheme that keep backwards compatible with @danielnielsennumber1 production usage, while at the same time extend the scheme with 0.0.0.0/0 matching for un-matched traffic.

Guess, xdp-cpumap-tc project now supports IPv6 :-)
